### PR TITLE
Teach Nexus about the Pantry

### PR DESCRIPTION
**Pull request by jmpesp**
Boot the new Pantry service zone, and use that to snapshot disks that are not attached to a currently running propolis. This functionality has to be driven in Nexus as part of a Saga because using the Pantry requires multiple steps (each one of which can fail). If this was simply a matter of proxying a call, the sled agent would still be used.

A simulated Pantry was implemented for the unit tests, and the logic for simulating a snapshot of an unattached disk moved there. The server code and Pantry interface were wholesale copied from the Crucible repo - it'd be nice to have a way to use the same interface(s) here and in Crucible (for the simulated agent too!) but I don't know a good way to share these cross-repo.

When the sled-agent brings up service zones, it inserts corresponding DNS records into the AZ's internal DNS server, and creating a Pantry client (or any service zone client) will need to query this internal DNS server for those records. This means that the simulated sled agent needs to also start an internal DNS server and insert records into it, which this commit implements. There's some minor refactoring in the internal-dns crate to create a `start` method for this purpose.

This means that Nexus needs to store the Resolver, and needs a way to update the Resoler that is used during integration tests. Importantly, services launched as part of integration tests will not use a standard known port (otherwise there would be many collisions), and a way to resolve both SRV (to get the port) and AAAA (to get the IP) records is required.

This commit also bumps the crucible and propolis git revs in the respective Cargo.toml files to the latest.

### CODE REVIEW

File: nexus/Cargo.toml [hunk 1]
```diff
@@ -47,7 +48,7 @@ oximeter-db = { path = "../oximeter/db/" }
 parse-display = "0.6.0"
 # See omicron-rpaths for more about the "pq-sys" dependency.
 pq-sys = "*"
-propolis-client = { git = "https://github.com/oxidecomputer/propolis", rev = "e58937fea67aca6bc2d37a3cc50a72291328ae0e", features = [ "generated" ] }
```
Comment (jmpesp) [inline nexus/Cargo.toml:50]: Note propolis#241 needs to be merged, then the propolis rev in this PR plus the relevant package-manifest bits need to be updated, before this PR goes in!

Comment (jmpesp) [reply nexus/Cargo.toml:50]: https://github.com/oxidecomputer/propolis/pull/241

File: nexus/src/app/sagas/snapshot_create.rs [hunk 2]
```diff
@@ -558,52 +654,374 @@ async fn ssc_send_snapshot_request(
...
+                    .map_err(ActionError::action_failed)?;
+            } else {
+                info!(
+                    log,
+                    "undo: disk {} not attached to pantry, it's attached to {}!",
+                    params.disk_id,
+                    attach_instance_id,
+                );
+
+                // XXX what to do here?
```
Comment (jmpesp) [inline nexus/src/app/sagas/snapshot_create.rs:745]: I'm really not sure what to do here. I'm also not sure if this situation can ever be reached!

The forward running of this saga should have "attached" the disk to the pantry. If that failed, then this undo node wouldn't be executed during the unwind. If that node succeeded, then the disk is "attached" to the pantry. The only way we would end up in this branch is if something else then took the already attached disk and attached it somewhere else, which seems like a violation of some contract somewhere :)

File: nexus/src/app/sagas/snapshot_create.rs [hunk 3]
```diff
@@ -558,52 +654,374 @@ async fn ssc_send_snapshot_request(
...
+                    .map_err(ActionError::action_failed)?;
+            } else {
+                info!(
+                    log,
+                    "disk {} not attached to pantry, it's attached to {}!",
+                    params.disk_id,
+                    attach_instance_id,
+                );
+
+                // XXX what to do here?
```
Comment (jmpesp) [inline nexus/src/app/sagas/snapshot_create.rs:1019]: Similarly, I don't know what to do here, and I'm not sure if this could ever really be reached.

At this point, the saga node that "attaches" this disk to the Pantry has succeeded, and all Pantry operations have succeeded. Again, to enter this branch, something must have come along, taken an already attached disk, and attached it somewhere else.

File: nexus/src/app/mod.rs [hunk 4]
```diff
@@ -79,7 +80,7 @@ pub struct Nexus {
     sec_client: Arc<steno::SecClient>,
 
     /// Task representing completion of recovered Sagas
-    recovery_task: std::sync::Mutex<Option<db::RecoveryTask>>,
+    recovery_task: Mutex<Option<db::RecoveryTask>>,
```
Comment (leftwo) [inline nexus/src/app/mod.rs:83]: Curious, why remove the `std::sync` here?  Just misc cleanup, or does this have some other impact?

**Review by leftwo — COMMENTED** [review]
To use the pantry, I'll need to take a snapshot of a disk that is not attached to a running propolis, right?

Is the pantry service always running on the sled agent, or do we create one on demand?

File: nexus/src/app/sagas/snapshot_create.rs [hunk 5]
```diff
@@ -145,9 +158,38 @@ lazy_static! {
...
         ssc_create_snapshot_record_undo,
     );
-    static ref SEND_SNAPSHOT_REQUEST: NexusAction = new_action_noop_undo(
-        "snapshot-create.send-snapshot-request",
-        ssc_send_snapshot_request,
+    static ref SEND_SNAPSHOT_REQUEST_TO_SLED_AGENT: NexusAction =
+        new_action_noop_undo(
+            "snapshot-create.send-snapshot-request-to-sled-agent",
+            ssc_send_snapshot_request_to_sled_agent,
+        );
```
Comment (leftwo) [inline nexus/src/app/sagas/snapshot_create.rs:165]: If the saga fails after this point, will this leave behind a snapshot?

File: nexus/src/app/sagas/snapshot_create.rs (hunk 2 above)
Comment (leftwo) [reply nexus/src/app/sagas/snapshot_create.rs:745]: Maybe we log a message and return error?  If we get here, then we need higher level help to unwind things.

File: nexus/src/app/disk.rs [hunk 6]
```diff
@@ -404,20 +405,61 @@ impl super::Nexus {
...
                 .organization_name(organization_name)
                 .project_name(project_name)
                 .disk_name(&db::model::Name(params.disk.clone()))
-                .lookup_for(authz::Action::Read)
+                .fetch_for(authz::Action::Read)
                 .await?;
 
+        // If there isn't a running propolis, Nexus needs to use the Crucible
+        // Pantry to make this snapshot
+        let use_the_pantry = if let Some(attach_instance_id) =
```
Comment (smklein) [inline nexus/src/app/disk.rs:420]: I'm seeing that we make the call on whether or not to use the pantry *super* early - like, before the saga even starts. I think that might run into some issue due to concurrent requests.

The current structure of this code is roughly:

1. Read Disk state from CRDB to see if there is an attached instance ID.
2. Use the output from (1) when constructing the saga DAG.
3. Run the actual saga.

What if the following ordering happens:

1. Read Disk state from CRDB to see if there is an attached instance ID. Let's pretend there *is* an attached disk.
**Let's suppose a concurrent request arrives now, and stops + destroys the instance** 
2. Use the output from (1) when constructing the saga DAG. The structure of the DAG assumes that it can contact the instance, and doesn't use the pantry.
3. Run the actual saga. See that the instance does not exist. Fail?

Would there be a problem with *unconditionally* sending the request to the pantry?

File: nexus/src/app/sagas/snapshot_create.rs (hunk 2 above)
Comment (leftwo) [reply nexus/src/app/sagas/snapshot_create.rs:734]: I added a message, but since we are in the undo path, this could be an expected situation.

File: nexus/src/app/sagas/snapshot_create.rs (hunk 5 above)
Comment (leftwo) [reply nexus/src/app/sagas/snapshot_create.rs]: Between work from @smklein and new undo steps added as part of the current rev, there should be no more
left behind snapshots.

File: nexus/src/app/disk.rs (hunk 6 above)
Comment (leftwo) [reply nexus/src/app/disk.rs:420]: I believe the above situation should result in the snapshot create saga failing.
I've added some more return `ActionError` when the snapshot create saga finds things different
than it expects, and I think returning error if things have changed under the snapshot is the
"right" thing to do here at this time.

File: nexus/src/app/mod.rs [hunk 7]
```diff
@@ -105,7 +106,10 @@ pub struct Nexus {
     // this amount of time is called "max issue delay" and we have to set that
     // in order for our integration tests that POST static SAML responses to
     // Nexus to not all fail.
-    samael_max_issue_delay: std::sync::Mutex<Option<chrono::Duration>>,
+    samael_max_issue_delay: Mutex<Option<chrono::Duration>>,
+
+    pub resolver:
+        Arc<tokio::sync::Mutex<internal_dns_client::multiclient::Resolver>>,
```
Comment (smklein) [inline nexus/src/app/mod.rs:112]: IMO it's questionable to expose a `pub` Mutex; this is easy to use incorrectly

File: nexus/src/app/mod.rs [hunk 8]
```diff
@@ -465,6 +473,20 @@ impl Nexus {
...
+    ) {
+        *self.resolver.lock().await = resolver;
+    }
+
+    pub async fn resolver(
+        &self,
+    ) -> tokio::sync::MutexGuard<'_, internal_dns_client::multiclient::Resolver>
+    {
+        self.resolver.lock().await
+    }
```
Comment (smklein) [inline nexus/src/app/mod.rs:489]: I think it's sketchy to expose a whole `MutexGuard` here. If any Nexus methods use the resolver, and we expose this functionality, a caller can do the following:

```rust
let resolver = nexus.resolver().await;
nexus.operation_which_uses_resolver().await; // Now we deadlock
```

File: nexus/src/app/sagas/snapshot_create.rs [hunk 9]
```diff
@@ -142,10 +155,32 @@ declare_saga_actions! {
...
+        + ssc_send_snapshot_request_to_sled_agent
+        - ssc_send_snapshot_request_to_sled_agent_undo
+    }
+    ATTACH_DISK_TO_PANTRY -> "attach_disk_to_pantry" {
+        + ssc_attach_disk_to_pantry
+        - ssc_attach_disk_to_pantry_undo
+    }
+    GET_PANTRY_ADDRESS -> "pantry_address" {
+        + ssc_get_pantry_address
+    }
```
Comment (smklein) [inline nexus/src/app/sagas/snapshot_create.rs:168]: I don't think we should embed the IP addresses of services in a saga - this can make it impossible to perform undo operations if a service is rebalanced.

For example:

- Action 1: Get pantry address
- Action 2: Send operation to pantry, using that address. Allocates a "Foo" object.
- Suppose service rebalancing happens, and the pantry moves to a new sled
- Action 3: Send another operation to pantry, using the old address. This fails.
- Undo Action 2: Cannot undo the "Foo" object allocation, the old pantry address no longer exists, and never will.

IMO we should be uncovering the pantry address as we use it.

File: nexus/src/app/sagas/snapshot_create.rs [hunk 10]
```diff
@@ -204,8 +239,32 @@ impl NexusSaga for SagaSnapshotCreate {
...
+            // (Pantry) Call the Pantry's /attach
+            builder.append(call_pantry_attach_for_disk_action());
+
+            // (Pantry) Call the Pantry's /snapshot
+            builder.append(call_pantry_snapshot_for_disk_action());
+
+            // (Pantry) Call the Pantry's /detach
+            builder.append(call_pantry_detach_for_disk_action());
+
+            // (Pantry) Clear attach_instance_id ZZZ verify, maybe DB too?
```
Comment (smklein) [inline nexus/src/app/sagas/snapshot_create.rs:264]: ZZZ?

File: nexus/src/app/disk.rs (hunk 6 above)
Comment (smklein) [reply nexus/src/app/disk.rs:420]: This still seems really sketchy to me. We explicitly throw an error for instances that are in unexpected states (see: `cannot snapshot attached disk for instance in state...`), but we don't really have a great way to validate that these sorts of transitions didn't happen immediately after our check. 

What about the other way? What if we didn't have an instance (so we expected to use the pantry), but we spin one up concurrently between performing this check and doing the saga? Would it be a problem to have the disk attached to the pantry at the same time as it's attached to an instance?

File: smf/sled-agent/config-rss.toml [hunk 11]
```diff
@@ -29,3 +29,11 @@ nexus_external_address = "192.168.1.20"
...
 # how-to-run.adoc for details on how to determine the value for your network.
 mac = "00:0d:b9:54:fe:e4"
+
+# [[request.service_zone]]
+# id = "1fa25cf1-147b-4179-8f8b-c601333733bf"
+# zone_name = "crucible-pantry"
+# addresses = [ "fd00:1122:3344:0101::10" ]
+# gz_addresses = []
+# [[request.service_zone.services]]
+# type = "crucible_pantry"
```
[the remark this answers is not in the document]
Comment (leftwo) [reply smf/sled-agent/config-rss.toml]: Yeah, leftover stuff I forgot to remove, fixed in 2428c0494e5a53f8fbf751b90b15bae1094edbe4

File: nexus/types/src/internal_api/params.rs [hunk 12]
```diff
@@ -128,11 +129,33 @@ impl fmt::Display for ServiceKind {
...
+            "dendrite" => Ok(Dendrite),
+            "tfport" => Ok(Tfport),
+            "crucible_pantry" => Ok(CruciblePantry),
+            _ => Err(Self::Err::InternalError {
+                internal_message: format!("Unknown service kind: {}", s),
+            }),
+        }
+    }
+}
+*/
```
[the remark this answers is not in the document]
Comment (leftwo) [reply nexus/types/src/internal_api/params.rs]: Yeah, leftover stuff I forgot to remove, fixed in 2428c0494e5a53f8fbf751b90b15bae1094edbe4

File: nexus/src/app/sagas/snapshot_create.rs (hunk 10 above)
Comment (leftwo) [reply nexus/src/app/sagas/snapshot_create.rs]: Yeah, more leftover stuff I forgot to remove, fixed in 2428c0494e5a53f8fbf751b90b15bae1094edbe4

File: nexus/src/app/disk.rs (hunk 6 above)
Comment (leftwo) [reply nexus/src/app/disk.rs:420]: I believe I now get what you are worried about here.

Your first `what if` was the right idea, but you picked a path where things would fail gracefully. 
It can indeed fail if:

1: The disk is not attached to an instance at the time of check, (pantry = true) 
  - After that check a new instance is created and the disk is attached to it.
  - Pantry will then steal the disk away from the instance.

2: Disk attached to a stopped instance at the time of check. (pantry = true)
  - After the check, the instance is started.
  - Pantry will steal the disk away from the instance.

I was able to get the pantry (with some well placed sleeps) to steal a disk away from an instance by doing 1 above.

I'm going to have to ponder this morass for a bit.

Ideally, either the snapshot can run to completion, or a change anywhere in the middle will cause the
snapshot to abort and cleanup behind itself.

File: internal-dns-client/src/multiclient.rs [hunk 13]
```diff
@@ -232,6 +236,39 @@ impl Resolver {
         Ok(*address)
     }
 
+    /// Looks up a single [`SocketAddrV6`] based on the SRV name
+    /// Returns an error if the record does not exist.
+    pub async fn lookup_socket_v6(
+        &self,
+        srv: crate::names::SRV,
+    ) -> Result<SocketAddrV6, ResolveError> {
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply internal-dns-client/src/multiclient.rs:244]: For the tests - if multiple simulated Pantrys are trying to use that port, there will be collisions. The simulated Pantry uses port `0` and then whatever port is chosen is inserted as part of a simulated SRV record.

File: internal-dns-client/src/multiclient.rs [hunk 14]
```diff
@@ -186,7 +190,7 @@ impl Resolver {
         Self::new_from_addrs(dns_addrs)
     }
 
-    fn new_from_addrs(
+    pub fn new_from_addrs(
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply internal-dns-client/src/multiclient.rs]: see b95ed918

File: nexus/src/app/mod.rs (hunk 7 above)
Comment (jmpesp) [reply nexus/src/app/mod.rs]: yeah, fixed in 1bd3ff98

File: nexus/src/app/mod.rs (hunk 8 above)
Comment (jmpesp) [reply nexus/src/app/mod.rs]: agreed, fixed in 1bd3ff98

File: nexus/src/app/sagas/snapshot_create.rs (hunk 9 above)
Comment (jmpesp) [reply nexus/src/app/sagas/snapshot_create.rs:168]: Unfortunately, the Pantry is not a stateless service that can be rebalanced like this. It will hold an activated Volume in memory and all requests referencing that Volume will need to be sent to the same instance of the service, hence the `GET_PANTRY_ADDRESS`.

If the Pantry crashes then it will lose all that state. If it restarts, further requests referencing that Volume will return errors, and sending those requests to other Pantrys will also return errors.

File: nexus/src/app/disk.rs (hunk 6 above)
Comment (jmpesp) [reply nexus/src/app/disk.rs:417]: 84825d09 adds a check for this disk attach race, let me know what you think

**Comment by jmpesp** [comment]
@smklein this is ready for a re-review :)

**Comment by jmpesp** [comment]
After working a bit on the "import to an image" prototype PR, I'm not happy about reusing the attached instance ID with the Pantry sentinel ID. I'm going to be adding a new `DiskState` called `Maintenance` specifically for this.

**Comment by jmpesp** [comment]
I ran into a lot of roadblocks adding the new `Maintenance` state I had in mind:

- I tried adding the variant `Maintenance { service_socket: SocketAddrV6 }` (and at the same time changing the other tuple variants in `DiskState` to struct variants), but OpenAPI doesn't support fields with this format. It does support fields with format ipv6 but then I'd have to embed all the fields of `SocketAddrV6` in. So I thought to use `#[serde(flatten)]` but unfortunately [it doesn't work for enums](https://github.com/serde-rs/serde/issues/1356). It also doesn't work with our DB schema because there isn't a native SocketAddrV6 type there either.

- DiskState in the database is a combination of a label and a uuid, which is combined using [a `impl TryFrom<(&str, Option<Uuid>)>`](https://github.com/oxidecomputer/omicron/blob/542970f7702559611938ffc59d2cb7e23787011e/common/src/api/external/mod.rs#L979) to create an `external::DiskState`. This isn't applicable to the maintenance variant because I wanted to store the SocketAddrV6.

- If I can't store the SocketAddrV6, then I can store the UUID of the selected Pantry service, and look up the SocketAddrV6 afterwards right? Nope! Our service schema [doesn't store port](https://github.com/oxidecomputer/omicron/blob/542970f7702559611938ffc59d2cb7e23787011e/nexus/db-model/src/service.rs#L20). The SRV record returned when querying for an available Pantry contains ipv6 and port, and the AAAA record it points to has the service's UUID. I've have to string split to get the service's UUID out of the front of it. This is straight forward but fragile.

So I changed it to not include any state. It's now just a `Maintenance` label, and the associated Pantry SocketAddrV6 is stored in saga state. This isn't the best but it beats all the work required to inline the DiskState enum with arbitrary fields in struct variants.

See ffc24c9a.

**Review by smklein — COMMENTED** [review]
If we're okay with the new API state, I think this validation makes sense. The gen-number-which-encourages-retry-amid-concurrent-ops helps a lot.

File: internal-dns-client/src/multiclient.rs [hunk 15]
```diff
@@ -232,6 +247,39 @@ impl Resolver {
...
+                let address = response.iter().next().ok_or_else(|| {
+                    ResolveError::NotFoundByString(name.to_string())
+                })?;
+
+                SocketAddrV6::new(*address, srv.port(), 0, 0)
+            }
+
+            _ => {
+                panic!("SRV lookup didn't return SRV!");
+            }
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply internal-dns-client/src/multiclient.rs]: agreed, b119eb43

File: internal-dns-client/src/multiclient.rs [hunk 16]
```diff
@@ -53,6 +54,16 @@ impl DnsAddressLookup for Ipv6Subnet<AZ_PREFIX> {
...
 
+impl DnsAddressLookup for SocketAddr {
+    fn dropshot_server_addrs(&self) -> Vec<SocketAddr> {
+        vec![*self]
+    }
+
+    fn dns_server_addrs(&self) -> Vec<SocketAddr> {
+        vec![*self]
+    }
+}
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply internal-dns-client/src/multiclient.rs]: done in 77bd516f, let me know

File: nexus/src/app/sagas/snapshot_create.rs [hunk 17]
```diff
@@ -697,6 +1046,76 @@ async fn ssc_send_snapshot_request_undo(
...
+                .map_err(ActionError::action_failed)?;
+        }
+
+        external::DiskState::Detached => {
+            info!(log, "disk {} already detached", params.disk_id);
+        }
+
+        _ => {
+            warn!(log, "disk is in state {:?}", db_disk.state());
+        }
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply nexus/src/app/sagas/snapshot_create.rs:1118]: If two concurrent operations are trying to bump a disk's generation number from N to N + 1, then only one can pass. If we're at this node in the saga, then this operation succeeded and the history of this disk's generation number and state is as follows:

```
Generation number | State
------------------|------------
            N + 0 | Detached
            N + 1 | Maintenance
```

Any attempt to bump the generation number of the disk from N + 1 to N + 2 will succeed if there are no other checks, and we would see two concurrent operations on the same disk. To make this situation impossible, no operation should begin on any disk in the maintenance state.

In the face of saga retries, let's say that this saga node executed fully but crashed before it could be committed to the database as finished. Then, the current state of the disk is "Detached" with the generation number N + 2. When this saga node is re-executed, there are two possibilities:

1. The disk is still "Detached" at generation number N + 2
2. The disk is in some unknown state at generation number N + 3

With 2, another operation started *after* the saga node executed but crashed, and *before* it was re-executed. This is a bug if there are further saga nodes to execute in the saga, which in this case there are. This call to `disk_update_runtime` has to be the last saga node that gets executed, which is done in 4c81a57e.

File: common/src/api/external/mod.rs [hunk 18]
```diff
@@ -919,6 +919,8 @@ pub enum DiskState {
     Creating,
     /// Disk is ready but detached from any Instance
     Detached,
+    /// Disk is undergoing maintenance
```
[the remark this answers is not in the document]
Comment (jmpesp) [reply common/src/api/external/mod.rs:961]: To summarize the chat: I'd like first change this label from `Maintenance` to something more specific, like `Taking a Snapshot`, and would like to decouple the internal `DiskState` from the view the user has. The user should only see `Busy` with some `String` reason, so their code can match on `DiskState::Busy(_)` instead of requiring an update every time we add a new maintenance-like operation like this (where the disk needs to be attached to the Pantry).

I'd love to do this in a follow up PR though, this one's been open for a long time and keeps accumulating merge conflicts.

File: nexus/src/app/disk.rs (hunk 6 above)
Comment (jmpesp) [reply nexus/src/app/disk.rs:417]: > Is it a desirable property that "attaching or detaching a disk to an instance" causes a concurrent snapshot request to fail? I'm trying to wrap my head around "should we be handling the snapshot request gracefully either way, or should we force a failure if this edge condition happens?"

To elaborate from our chat: our current implementation can't support concurrent operations where a disk's volume is being activated by multiple Upstairs, so we have to cause these concurrent requests to fail.

> It seems like the disk still exists in either state, so it really could be snapshottable at any time, right?

From the perspective of an ideal cloud, a disk should be snapshottable at any time, yeah :) We may support this in the future but certainly not today.

> In the case where we want to throw an error, are we propagating the correct error back to the user? We should tell them to retry, but I suspect a string-based ActionError::action_failed error translates to a 500 error.

Yeah, this is likely a 500. I'm not aware of any HTTP response code that fits nicely with "this will work if you retry it but can't work now". The user sent a well-formed request but the server was unable to process it, so 500 kinda fits.

A 503 might fit here? Problem is user code might not interpret a 503 as "please try again" :)

> Are we testing this anywhere?

To test this, we would need machinery in our integration test framework to construct two sagas, and execute individual nodes from those sagas manually.

File: common/src/api/external/mod.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (smklein) [reply common/src/api/external/mod.rs:961]: Sounds good. If we're aligned on long-term intent, I'm happy to merge this earlier and do follow-up work in separate PRs.

File: internal-dns-client/src/multiclient.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (smklein) [reply internal-dns-client/src/multiclient.rs]: Got it, I'm happy with this approach. As a nit, can we add a docstring comment to `ServerAddresses`, and perhaps giving it a more test-specific name, to imply that "this is used to control addresses during testing and avoid collisions -- normally, the DNS addresses are inferred from the underlay"

File: nexus/src/app/sagas/snapshot_create.rs [hunk 19]
```diff
@@ -697,6 +1051,76 @@ async fn ssc_send_snapshot_request_undo(
...
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., authz_disk, db_disk) =
+        LookupPath::new(&opctx, &osagactx.datastore())
+            .disk_id(params.disk_id)
+            .fetch_for(authz::Action::Modify)
+            .await
+            .map_err(ActionError::action_failed)?;
```
Comment (smklein) [inline nexus/src/app/sagas/snapshot_create.rs:1091]: I think it's possible that this node might cause a data corruption in a case where we are concurrently snapshotting the same disk.

- Suppose we have a disk, D.
- Suppose we are executing two sagas on different nexus nodes to snapshot the disk D. 
- Saga 1 executes *almost* entirely, getting through this node, then crashing. This leaves the disk in state: "Detached", but `ssc_detach_disk_from_pantry` is not yet marked as done.
- Saga 2 starts, and is mid-way through the snapshot operation. The state of the disk is "Maintenance".
- Saga 1 restarts on a rebooted nexus node. It retries `ssc_detach_disk_from_pantry`.

With the current implementation, I think this would cause "saga 1" to pull the disk *out* of maintenance, even though it's actually being used by "saga 2".

I think the crux of this problem is that the generation accounting on the disk is lost, because the `db_disk` value is re-read from CRDB. As you mentioned [here](https://github.com/oxidecomputer/omicron/pull/1916#discussion_r1095108519), we would *want* to proceed with the "detach" operation only if the generation number is currently equal our expected value.

In other words, I think this is what we *want* to do during a snapshot saga using the pantry:
| Generation Number | State |
| --- | --- |
| N | Detached |
| N + 1 | Maintenance - but we only move it here if the prior `Generation` was N |
| N + 2 | Detached - but we only move it here if the prior `Generation` was N + 1 |

But I think this is what we're actually doing:
In other words, I think this is what we *want* to do during a snapshot saga using the pantry:
| Generation Number | State |
| --- | --- |
| N | Detached |
| N + 1 | Maintenance - but we only move it here if the prior `Generation` was N |
| X | Detached - but we only move it here if the prior `State` was `Maintenance`. But it might be in maintenance because of an operation made by a different saga! |

Comment (jmpesp) [reply nexus/src/app/sagas/snapshot_create.rs:1101]: I think you're right, yeah. 7ff4beb4 records the generation number N+1 (after setting to Maintenance), then looks it up later to verify that expectation before setting to Detached

**Review by smklein — APPROVED** [review]
Thanks for the iteration, I appreciate you taking the time to handle these edge cases.

I still think a test or too covering the behavior mentioned here (https://github.com/oxidecomputer/omicron/pull/1916#discussion_r1095961835) would be useful, but your call on when / where to merge that.

File: nexus/src/app/sagas/snapshot_create.rs [hunk 20]
```diff
@@ -1987,4 +2003,255 @@ mod test {
             dag = create_saga_dag::<SagaSnapshotCreate>(params).unwrap();
         }
     }
+
+    #[nexus_test(server = crate::Server)]
```
Comment (smklein) [inline nexus/src/app/sagas/snapshot_create.rs:2007]: Thanks for the tests, they look great!

**Comment by jmpesp** [comment]
the `helios / deploy` failure is caused by an out-dated crucible dependency here, which requires bumping propolis' crucible revs, then both propolis and crucible revs here.

**Review by leftwo — APPROVED** [review]
The one question I had about:
 * Disk attached to stopped instance
 * Snapshot request SAGA starts -> use pantry is true so pantry spins up.
 * Instance is started while snapshot is underway.

This should return a 500 error, which is what we want.

**Comment by smklein** [comment]
> The one question I had about:
> 
> * Disk attached to stopped instance
> * Snapshot request SAGA starts -> use pantry is true so pantry spins up.
> * Instance is started while snapshot is underway.
> 
> This should return a 500 error, which is what we want.

I am not sure I agree that this should be a 500 error.

See: https://github.com/oxidecomputer/omicron/blob/main/docs/http-status-codes.adoc - any time that a 500 error is returned, **we should treat it as a bug**. In other words, this is a condition in which someone should be filing a bug.

Since this is predictable behavior, which I'm assuming we're *intending* for this PR, we should return 503 to imply "the client should retry the operation".
